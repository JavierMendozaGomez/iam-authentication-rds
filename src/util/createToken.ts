
import {RDS} from 'aws-sdk';
import * as util from 'util';

const createToken = async (): Promise<string> => {
    const signer = new RDS.Signer({
        region: process.env.DB_REGION,
        hostname: process.env.DB_HOSTNAME,
        port: parseInt(process.env.DB_PORT),
        username: process.env.DB_USER,
    });
    const getAuthTokenAsync = util.promisify(signer.getAuthToken).bind(signer);
    return await getAuthTokenAsync({username: process.env.DB_AURORA_USER});
};

export default createToken;
