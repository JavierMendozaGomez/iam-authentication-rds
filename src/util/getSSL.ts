import {promisify} from 'util';
import * as fs from 'fs';

const getSSL = async () => {
    const readFileAsync = promisify(fs.readFile);
    return await readFileAsync(`${__dirname}/rds-ca-2015-eu-west-2.pem`);
};

export default getSSL;
