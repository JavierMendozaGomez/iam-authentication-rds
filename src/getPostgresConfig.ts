import createToken from './util/createToken';
import getSSL from './util/getSSL';
import {ClientConfig} from 'pg';

const getPostgresConnectionConfig = async (): Promise<ClientConfig> => {
    const ssl = await getSSL();
    const token = await createToken();
    return {
        host: process.env.DB_HOSTNAME,
        port: parseInt(process.env.DB_PORT),
        user: process.env.DB_USER,
        password: token,
        database: process.env.DB_NAME,
        ssl: {
            ca: ssl,
        },
    };
};

export default getPostgresConnectionConfig;
