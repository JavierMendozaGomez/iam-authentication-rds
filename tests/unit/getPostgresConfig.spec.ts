import  getPostgresConfig from '../../src/getPostgresConfig';
import createToken from '../../src/util/createToken';
import getSSL from '../../src/util/getSSL';
jest.mock('../../src/util/createToken');
jest.mock('../../src/util/getSSL');

const db = {
    DB_REGION: process.env.DB_REGION,
    DB_HOSTNAME: process.env.DB_HOSTNAME,
    DB_PORT: process.env.DB_PORT,
    DB_USER: process.env.DB_USER,
    DB_NAME: process.env.DB_NAME,
};

describe('Unit test getPostgresConfig', () => {
    beforeAll(() => {
        process.env.DB_REGION = 'mockedRegion';
        process.env.DB_HOSTNAME = 'mockedHostname';
        process.env.DB_PORT = '123456';
        process.env.DB_USER = 'mockedDBUser';
        process.env.DB_NAME = 'mockedDBName';
    });

    afterAll(() => {
        process.env.DB_REGION = db.DB_REGION;
        process.env.DB_HOSTNAME = db.DB_REGION;
        process.env.DB_PORT = db.DB_PORT;
        process.env.DB_USER = db.DB_USER;
        process.env.DB_NAME = db.DB_NAME;
    });

    it(`Successfully gets the info to connect the db`, async () => {
        //@ts-ignore
        createToken.mockResolvedValue('DummyToken');
        //@ts-ignore
        getSSL.mockResolvedValue('11 22 33 44');
        expect(await getPostgresConfig()).toMatchSnapshot();
    });
});
