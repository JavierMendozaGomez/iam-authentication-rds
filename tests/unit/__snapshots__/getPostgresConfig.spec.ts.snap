// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit test getPostgresConfig Successfully gets the info to connect the db 1`] = `
Object {
  "database": "mockedDBName",
  "host": "mockedHostname",
  "password": "DummyToken",
  "port": 123456,
  "ssl": Object {
    "ca": "11 22 33 44",
  },
  "user": "mockedDBUser",
}
`;
